cms/djangoapps/contentstore/outlines.py

```diff
@@ -0,0 +1,201 @@
+learning_sequences at publish time.
+"""
+from datetime import timezone
+
+from edx_django_utils.monitoring import function_trace, set_custom_attribute
+
+from openedx.core.djangoapps.content.learning_sequences.api import replace_course_outline
+from openedx.core.djangoapps.content.learning_sequences.data import (
+    CourseLearningSequenceData,
+    CourseOutlineData,
+    CourseSectionData,
+    CourseVisibility,
+    ExamData,
+    VisibilityData,
+)
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+
+
+class CourseStructureError(Exception):
+    """
+    Raise this if we can't create an outline because of the course structure.
+
+    Courses built in Studio conform to a hierarchy that looks like:
+        Course -> Section (a.k.a. Chapter) -> Subsection (a.k.a. Sequence)
+
+    OLX imports are much more freeform and can generate unusual structures that
+    we won't know how to handle.
+    """
+
+
+def _check_sequence_fields(sequence):
+    """
+    Raise CourseStructureError if `sequence` is missing a required field.
+
+    Do this instead of checking against specific block types to better future
+    proof ourselves against new sequence-types, aliases, changes in the way
+    dynamic mixing of XBlock types happens, as well as deprecation/removal of
+    the specific fields we care about. If it quacks like a duck...
+    """
+    expected_fields = [
+        'display_name',
+        'hide_after_due',
+        'hide_from_toc',
+        'is_practice_exam',
+        'is_proctored_enabled',
+        'is_time_limited',
+        'visible_to_staff_only',
+    ]
+    for field in expected_fields:
+        if not hasattr(sequence, field):
+            msg = (
+                f"Cannot create CourseOutline: Expected a Sequence at "
+                f"{sequence.location} (child of {sequence.parent}), "
+                f"but this object does not have sequence field {field}."
+            )
+            raise CourseStructureError(msg)
+
+
+def _check_section_fields(section):
+    """
+    Raise CourseStructureError if `section` is missing a required field.
+
+    Do this instead of checking against specific block types to better future
+    proof ourselves against new sequence-types, aliases, changes in the way
+    dynamic mixing of XBlock types happens, as well as deprecation/removal of
+    the specific fields we care about. If it quacks like a duck...
+    """
+    expected_fields = [
+        'children',
+        'hide_from_toc',
+        'visible_to_staff_only',
+    ]
+    for field in expected_fields:
+        if not hasattr(section, field):
+            msg = (
+                f"Cannot create CourseOutline: Expected a Section at "
+                f"{section.location} (child of {section.parent}), "
+                f"but this object does not have Section field {field}."
+            )
+            raise CourseStructureError(msg)
```

Comment:
@kdmccormick: Basic validation for course structure put here. Sample error message is:

`cms.djangoapps.contentstore.outlines.CourseStructureError: Cannot create CourseOutline: Expected a Sequence at block-v1:TNL+7733+OutlineFromModuleStoreTestCase+type@vertical+block@Unit (child of block-v1:TNL+7733+OutlineFromModuleStoreTestCase+type@sequential+block@Sequence), but this object does not have sequence field is_practice_exam.`

Reply:
:duck: :+1: